From bc7ed693f9e66685a80465ba41b9a4bbdee41672 Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
Date: Wed, 13 Dec 2023 18:30:51 +0530
Subject: [PATCH 2/2] wifi: ath12k: prevent KASAN warning in _find_next_bit

for_each_set_bit() invoking _find_next_bit expects only unsigned long arguments
which causes KASAN warning for u16 usage of links_map variable.

Fix this by declaring ahsta->links_map, ahvif->links_map as unsigned long
variables and modify debug prints accordingly.

Signed-off-by: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c  |  2 +-
 drivers/net/wireless/ath/ath12k/core.h  |  4 +--
 drivers/net/wireless/ath/ath12k/dp_rx.c |  4 +--
 drivers/net/wireless/ath/ath12k/dp_tx.c |  8 ++---
 drivers/net/wireless/ath/ath12k/mac.c   | 48 ++++++++-----------------
 drivers/net/wireless/ath/ath12k/peer.h  |  2 +-
 drivers/net/wireless/ath/ath12k/sawf.c  |  4 +--
 7 files changed, 25 insertions(+), 47 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 6cd03e906586..e1743e48ec62 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -2737,7 +2737,7 @@ static int ath12k_mlo_core_recovery_reconfig_link_bss(struct ath12k *ar,
 	ret = 0;
 exit:
 	ath12k_dbg(ab, ATH12K_DBG_MODE1_RECOVERY,
-		   "ret:%d No. of vdev created:%d, links_map:%d, flag:%d\n",
+		   "ret:%d No. of vdev created:%d, links_map:%lu, flag:%d\n",
 		   ret,
 		   ahvif->num_vdev_created,
 		   ahvif->links_map,
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index d2dee24f789d..ab6be0c6540d 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -479,7 +479,7 @@ struct ath12k_vif {
 	struct ath12k_link_vif deflink;
 	struct ath12k_link_vif *link[ATH12K_NUM_MAX_LINKS];
 	/* indicates bitmap of link vdev created in FW */
-	u16 links_map;
+	unsigned long links_map;
 	u32 aid;
 	struct ath12k_mgmt_frame_stats mgmt_stats;
 	u8 num_vdev_created;
@@ -704,7 +704,7 @@ struct ath12k_sta {
 	struct ath12k_link_sta deflink;
 	struct ath12k_link_sta *link[IEEE80211_MLD_MAX_NUM_LINKS];
 	/* indicates bitmap of link sta created in FW */
-	u16 links_map;
+	unsigned long links_map;
 	u16 ml_peer_id;
 	u8 assoc_link_id;
 	u8 num_peer;
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 3c32f976a52d..69b4e4638333 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -1106,13 +1106,11 @@ void ath12k_dp_tid_setup(void *data, struct ieee80211_sta *sta)
 	struct ath12k_link_sta *arsta;
 	struct ath12k_link_vif *arvif;
 	u8 link_id;
-	u16 links_map;
 
 	if (sta->mlo)
 		return;
 
-	links_map = ahsta->links_map;
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahsta->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arsta = ahsta->link[link_id];
 		if (!arsta)
 			continue;
diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 13321c7a0560..b634cc9746f1 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -880,7 +880,7 @@ ath12k_dp_tx_htt_tx_complete_buf(struct ath12k_base *ab,
 					ahvif->link[skb_cb->link_id]->link_stats.tx_completed++;
 			} else {
 				ath12k_warn(ab,
-					    "invalid linkid 0x%X in htt tx complete buf linkmap 0x%X\n",
+					    "invalid linkid 0x%X in htt tx complete buf linkmap %lu\n",
 					    skb_cb->link_id,
 					    ahvif->links_map);
 			}
@@ -906,7 +906,7 @@ ath12k_dp_tx_htt_tx_complete_buf(struct ath12k_base *ab,
 			ahvif->link[skb_cb->link_id]->link_stats.tx_completed++;
 	} else {
 		ath12k_warn(ab,
-			    "invalid linkid 0x%X in htt tx complete buf linkmap 0x%X\n",
+			    "invalid linkid 0x%X in htt tx complete buf linkmap %lu\n",
 			    skb_cb->link_id,
 			    ahvif->links_map);
 	}
@@ -1172,7 +1172,7 @@ static void ath12k_dp_tx_complete_msdu(struct ath12k *ar,
 			ahvif->link[skb_cb->link_id]->link_stats.tx_completed++;
 	} else {
 		ath12k_warn(ar->ab,
-			    "invalid linkid 0x%X in tx complete msdu linkmap 0x%X\n",
+			    "invalid linkid 0x%X in tx complete msdu linkmap %lu\n",
 			    skb_cb->link_id,
 			    ahvif->links_map);
 	}
@@ -1653,7 +1653,7 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 						ahvif->link[skb_cb->link_id]->link_stats.tx_completed++;
 				} else {
 					ath12k_warn(ab,
-						    "invalid linkid 0x%X in tx completion handler for linkmap 0x%X\n",
+						    "invalid linkid 0x%X in tx completion handler for linkmap %lu\n",
 						    skb_cb->link_id,
 						    ahvif->links_map);
 				}
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 21e09d226ff5..b527d16da804 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -818,16 +818,13 @@ static void ath12k_get_arvif_iter(void *data, u8 *mac,
 	struct ath12k_vif_iter *arvif_iter = data;
 	struct ath12k_vif *ahvif = (void *)vif->drv_priv;
 	struct ath12k_link_vif *arvif;
-	u16 links_map;
 	u8 link_id;
 
 	/* TODO Revisit lock usage for ahvif->link. Currently this is
 	 * used with vif protected by rcu lock and since the ahvif->links_map
 	 * or link wont be modified it is safe to access below without its mutex
 	 */
-	links_map = ahvif->links_map;
-
-	for_each_set_bit(link_id, &links_map, ATH12K_NUM_MAX_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arvif = ahvif->link[link_id];
 
 		if (WARN_ON(arvif == NULL))
@@ -4945,7 +4942,6 @@ static void ath12k_mac_op_vif_cfg_changed(struct ieee80211_hw *hw,
 	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
 	struct ath12k_link_vif *arvif;
 	struct ath12k *ar;
-	unsigned long links_map;
 	unsigned long link_id;
 	struct ieee80211_bss_conf *info;
 
@@ -4957,10 +4953,9 @@ static void ath12k_mac_op_vif_cfg_changed(struct ieee80211_hw *hw,
 	}
 
 	link_id = ahvif->deflink.link_id;
-	links_map = ahvif->links_map;
 
 	if (changed & BSS_CHANGED_ASSOC) {
-		for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+		for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 			arvif = ahvif->link[link_id];
 			info = vif->link_conf[link_id];
 			ar = arvif->ar;
@@ -6962,15 +6957,12 @@ static void ath12k_sta_set_4addr_wk(struct work_struct *wk)
 	struct ath12k_link_sta *arsta;
 	struct ieee80211_sta *sta;
 	int ret = 0;
-	u16 links_map;
 	u8 link_id;
 
 	ahsta = container_of(wk, struct ath12k_sta, set_4addr_wk);
 	sta = container_of((void *)ahsta, struct ieee80211_sta, drv_priv);
 
-	links_map = ahsta->links_map;
-
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahsta->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arsta = ahsta->link[link_id];
 		arvif = arsta->arvif;
 		ar = arvif->ar;
@@ -7752,6 +7744,7 @@ static int ath12k_mac_op_change_sta_links(struct ieee80211_hw *hw,
 	struct ath12k_link_sta *arsta;
 	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
 	struct ath12k_sta *ahsta = ath12k_sta_to_ahsta(sta);
+	unsigned long valid_links;
 	u8 link_id;
 	int ret = 0;
 
@@ -7775,7 +7768,8 @@ static int ath12k_mac_op_change_sta_links(struct ieee80211_hw *hw,
 	/* this op is expected only after initial sta insertion with default link */
 	WARN_ON(ahsta->links_map == 0);
 
-	for_each_set_bit(link_id, &new_links, IEEE80211_MLD_MAX_NUM_LINKS) {
+	valid_links = new_links;
+	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
 		if (ahsta->links_map & BIT(link_id))
 			continue;
 
@@ -9191,7 +9185,7 @@ static void ath12k_mgmt_over_wmi_tx_work(struct work_struct *work)
 		ahvif = ath12k_vif_to_ahvif(skb_cb->vif);
 		if (!(ahvif->links_map & BIT(skb_cb->link_id))) {
 			ath12k_warn(ar->ab,
-				    "invalid linkid 0x%X in mgmt over wmi tx with linkmap 0x%X\n",
+				    "invalid linkid 0x%X in mgmt over wmi tx with linkmap %lu\n",
 				    skb_cb->link_id,
 				    ahvif->links_map);
 			ath12k_mgmt_over_wmi_tx_drop(ar, skb);
@@ -9406,11 +9400,11 @@ static u8 ath12k_mac_get_tx_link(struct ieee80211_sta *sta, struct ieee80211_vif
 
 
 			ath12k_dbg(NULL, ATH12K_DBG_MAC,
-				   "unable to determine link sta addr for translation - Frame control %x link %d sta links %x ahsta links %x\n",
+				   "unable to determine link sta addr for translation - Frame control %x link %d sta links %x ahsta links %lu\n",
 				   hdr->frame_control, link, sta->valid_links,
 				   ahsta->links_map);
 		} else {
-			ath12k_err(NULL, "unable to determine Tx link for frame %x link %d\n vif links %x sta links %x ahvif links %x ahsta links %x",
+			ath12k_err(NULL, "unable to determine Tx link for frame %x link %d\n vif links %x sta links %x ahvif links %lu ahsta links %lu",
 				   hdr->frame_control, link, vif->valid_links,
 				   sta->valid_links, ahvif->links_map, ahsta->links_map);
 			link = IEEE80211_MLD_MAX_NUM_LINKS;
@@ -9676,7 +9670,7 @@ static void ath12k_mac_op_tx(struct ieee80211_hw *hw,
 			if (!peer) {
 				spin_unlock_bh(&tmp_ar->ab->base_lock);
 				ath12k_warn(tmp_ar->ab,
-					    "failed to find peer for vdev_id 0x%X addr %pM link_map 0x%X\n",
+					    "failed to find peer for vdev_id 0x%X addr %pM link_map %lu\n",
 					    tmp_arvif->vdev_id, tmp_arvif->addr,
 					    ahvif->links_map);
 				tmp_ar->ab->soc_stats.tx_err.peers_not_present++;
@@ -11243,15 +11237,12 @@ static void ath12k_mac_op_remove_interface(struct ieee80211_hw *hw,
 	struct ath12k_vif *ahvif = (void *)vif->drv_priv;
 	struct ath12k_link_vif *arvif;
 	struct ath12k *ar;
-	u16 links_map;
 	u8 link_id;
 	int ret;
 
 	mutex_lock(&ah->conf_mutex);
 
-	links_map = ahvif->links_map;
-	
-	for_each_set_bit(link_id, &links_map, ATH12K_NUM_MAX_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, ATH12K_NUM_MAX_LINKS) {
 		arvif = ahvif->link[link_id];
 
 		if (!arvif)
@@ -11573,7 +11564,6 @@ ath12k_mac_mlo_get_vdev_args(struct ath12k_link_vif *arvif,
 {
 	struct ath12k_vif *ahvif = arvif->ahvif;
 	struct ath12k_link_vif *arvif_p;
-	u16 links_map;
 	u8 link_id;
 	struct wmi_ml_partner_info *partner_info;
 	struct ieee80211_bss_conf *link_conf;
@@ -11596,11 +11586,9 @@ ath12k_mac_mlo_get_vdev_args(struct ath12k_link_vif *arvif,
 
 	/* TODO check assoc and mcast vdev for AP mode */
 
-	links_map = ahvif->links_map;
-
 	partner_info = ml_arg->partner_info;
 
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arvif_p = ahvif->link[link_id];
 
 		if (WARN_ON(arvif_p == NULL))
@@ -11989,16 +11977,13 @@ ath12k_mac_change_chanctx_cnt_iter(void *data, u8 *mac,
 	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
 	struct ath12k_link_vif *arvif;
 	struct ath12k_mac_change_chanctx_arg *arg = data;
-	u16 links_map;
 	u8 link_id;
 	struct ieee80211_bss_conf *link_conf;
 	struct ath12k_hw *ah = ahvif->ah;
 
 	lockdep_assert_held(&ah->conf_mutex);
 
-	links_map = ahvif->links_map;
-
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arvif = ahvif->link[link_id];
  
 		if (WARN_ON(arvif == NULL))
@@ -12025,16 +12010,13 @@ ath12k_mac_change_chanctx_fill_iter(void *data, u8 *mac,
 	struct ath12k_link_vif *arvif;
 	struct ath12k_mac_change_chanctx_arg *arg = data;
 	struct ieee80211_chanctx_conf *ctx;
-	u16 links_map;
 	u8 link_id;
 	struct ieee80211_bss_conf *link_conf;
 	struct ath12k_hw *ah = ahvif->ah;
 
 	lockdep_assert_held(&ah->conf_mutex);
 
-	links_map = ahvif->links_map;
- 
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arvif = ahvif->link[link_id];
 
 		if (WARN_ON(arvif == NULL))
@@ -13473,7 +13455,7 @@ ath12k_mac_op_unassign_vif_chanctx(struct ieee80211_hw *hw,
 	}
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
-		   "mac unassign vif links map 0x%x vdev id %d addr %pM assoc_link %d\n",
+		   "mac unassign vif links map %lu vdev id %d addr %pM assoc_link %d\n",
 		   ahvif->links_map, arvif->vdev_id, arvif->addr, arvif->assoc_link);
 
 	/* The primary link should be stop vdev and delete peer as last one. */
diff --git a/drivers/net/wireless/ath/ath12k/peer.h b/drivers/net/wireless/ath/ath12k/peer.h
index c3788cb2c676..4e06d97b890c 100644
--- a/drivers/net/wireless/ath/ath12k/peer.h
+++ b/drivers/net/wireless/ath/ath12k/peer.h
@@ -136,7 +136,7 @@ struct ath12k_link_sta *ath12k_peer_get_link_sta(struct ath12k_base *ab,
 	ahsta = ath12k_sta_to_ahsta(peer->sta);
 	if (peer->ml_peer_id & ATH12K_ML_PEER_ID_VALID) {
 		if (!(ahsta->links_map & BIT(peer->link_id))) {
-			ath12k_warn(ab, "peer %pM id %d link_id %d can't found in STA link_map 0x%x\n",
+			ath12k_warn(ab, "peer %pM id %d link_id %d can't found in STA link_map %lu\n",
 				    peer->addr, peer->peer_id, peer->link_id, ahsta->links_map);
 			return NULL;
 		}
diff --git a/drivers/net/wireless/ath/ath12k/sawf.c b/drivers/net/wireless/ath/ath12k/sawf.c
index e01fea9b4cbf..f5fb84e6b4ad 100644
--- a/drivers/net/wireless/ath/ath12k/sawf.c
+++ b/drivers/net/wireless/ath/ath12k/sawf.c
@@ -485,7 +485,6 @@ struct ath12k_base *ath12k_sawf_get_ab_from_netdev(struct net_device *dev,
 	struct ath12k_peer *peer;
 	struct ath12k_vif *ahvif;
 	struct ath12k_link_vif *arvif;
-	u16 links_map;
 	u8 link_id;
 
 	if (!dev)
@@ -506,10 +505,9 @@ struct ath12k_base *ath12k_sawf_get_ab_from_netdev(struct net_device *dev,
 	/* ToDo: Instead of mutex_lock, rcu locking will be used
 	   to access the ahvif->link
 	*/
-	links_map = ahvif->links_map;
 	/* Look into all the links, find the matching peer and
 	 * return with ab and peer_id detail */
-	for_each_set_bit(link_id, &links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
+	for_each_set_bit(link_id, &ahvif->links_map, IEEE80211_MLD_MAX_NUM_LINKS) {
 		arvif = ahvif->link[link_id];
 
 		if (WARN_ON(!arvif))
-- 
2.25.1

